desc/integrals/basis.py

```diff
@@ -58,6 +85,627 @@ def _in_epigraph_and(is_intersect, df_dy_sign, /):
+        # Add axis to broadcast against Chebyshev coefficients.
+        x = jnp.atleast_1d(x)[..., jnp.newaxis]
+        # Add axis to broadcast against multiple x values.
+        cheb = cheb_from_dct(
+            irfft_non_uniform(x, self._c[..., jnp.newaxis, :, :], self.M, axis=-2)
+        )
+        assert cheb.shape[-2:] == (x.shape[-2], self.N)
+        return PiecewiseChebyshevSeries(cheb, self.domain)
+
+
+class PiecewiseChebyshevSeries(IOAble):
+    """Chebyshev series.
+
+    { fₓ | fₓ : y ↦ ∑ₙ₌₀ᴺ⁻¹ aₙ(x) Tₙ(y) }
+    and Tₙ are Chebyshev polynomials on [−yₘᵢₙ, yₘₐₓ]
+
+    Attributes
+    ----------
+    cheb : jnp.ndarray
+        Shape (..., M, N).
+        Chebyshev coefficients αₙ(x) for fₓ(y) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x) Tₙ(y).
+    M : int
+        Number of cuts.
+    N : int
+        Chebyshev spectral resolution of each cut.
+    domain : (float, float)
+        Domain for y coordinates.
+
+    """
+
+    def __init__(self, cheb, domain=(-1, 1)):
+        """Make piecewise series from given Chebyshev coefficients.
+
+        Parameters
+        ----------
+        cheb : jnp.ndarray
+            Shape (..., M, N).
+            Chebyshev coefficients αₙ(x=``x``) for f(x, y) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x) Tₙ(y).
+        domain : (float, float)
+            Domain for y coordinates. Default is [-1, 1].
+
+        """
+        self.cheb = jnp.atleast_2d(cheb)
+        errorif(domain[0] > domain[-1], msg="Got inverted domain.")
+        self.domain = tuple(domain)
+
+    @property
+    def M(self):
+        """Number of cuts."""
+        return self.cheb.shape[-2]
+
+    @property
+    def N(self):
+        """Chebyshev spectral resolution."""
+        return self.cheb.shape[-1]
+
+    def evaluate(self, N):
+        """Evaluate Chebyshev series at N Chebyshev points.
+
+        Evaluate each function in this set
+        { fₓ | fₓ : y ↦ ∑ₙ₌₀ᴺ⁻¹ aₙ(x) Tₙ(y) }
+        at y points given by the N Chebyshev points.
+
+        Parameters
+        ----------
+        N : int
+            Grid resolution in y direction. Preferably power of 2.
+
+        Returns
+        -------
+        fq : jnp.ndarray
+            Shape (..., M, N)
+            Chebyshev series evaluated at N Chebyshev points.
+
+        """
+        return idct(dct_from_cheb(self.cheb), type=2, n=N, axis=-1) * N
+
+    def isomorphism_to_C1(self, y):
+        """Return coordinates z ∈ ℂ isomorphic to (x, y) ∈ ℂ².
+
+        Maps row x of y to z = y + f(x) where f(x) = x * |domain|.
+
+        Parameters
+        ----------
+        y : jnp.ndarray
+            Shape (..., y.shape[-2], y.shape[-1]).
+            Second to last axis iterates the rows.
+
+        Returns
+        -------
+        z : jnp.ndarray
+            Shape y.shape.
+            Isomorphic coordinates.
+
+        """
+        assert y.ndim >= 2
+        z_shift = jnp.arange(y.shape[-2]) * (self.domain[-1] - self.domain[0])
+        z = y + z_shift[:, jnp.newaxis]
+        return z
+
+    def isomorphism_to_C2(self, z):
```

Comment:
is this basically doing the field line wrapping from one alpha to another? Wouldn't that need to know about iota?

Reply:
That's what the function `get_alpha` does in `bounce_utils`. This is more about mapping z along field line to (i, z % 2pi) so that we can select which cut of the field line to evaluate on. Similar to [your code](https://github.com/f0uriest/interpax/blob/0072a47af1758ae56493971666d5de242d37f46d/interpax/_ppoly.py#L234), except without the unnecessary searching

Reply:
(It's unnecessary in this case because I know all the piecewise polynomials are defined over the same length, 0 to 2pi. You can't do that in the code I linked because your knots need not be uniformly spaced).